metrics: add build command duration metric

jsternberg opened the pull request:
This adds a build duration metric for the build command with attributes 
related to the buildx driver, the error type (if any), and which options 
were used to perform the build from a subset of the options.

crazy-max commented on util/osutil/path_windows.go:
```diff
@@ -1,10 +1,8 @@
-package build
+package osutil
 
-import "golang.org/x/sys/windows"
```
Looks missing

jsternberg replied:
Good catch. I'll have to watch for that in the future. I didn't realize my IDE wasn't running `goimports` on files that didn't match my build constraints.

tonistiigi commented on commands/build.go:
```diff
@@ -279,6 +362,7 @@ func runBuild(ctx context.Context, dockerCli command.Cli, options buildOptions)
 		return err
 	}
 
+	done := timeBuildCommand(mp, options.Attributes(dockerCli))
```
Make this `(build)metricsAttributes(b, dockerCli, options) attribute.Set`.

Don't need `options.newBuilder()` or other methods on `options`.

Inside the function, read the `b.Driver` directly from input without the struct/once. If you want then for the hash you can leave the lazy evaluation, but for just reading the driver property, there is no need for this overhead.

The call could also be inside `timeBuildCommand`.

jsternberg replied:
Made the change. This isn't called inside of `timeBuildCommand` as there's another section of code that's going to use these same attributes in another PR.

tonistiigi commented on commands/build.go:
```diff
@@ -930,3 +1016,38 @@ func maybeJSONArray(v string) []string {
+func timeBuildCommand(mp metric.MeterProvider, attrs attribute.Set) func(err error) {
+	meter := metricutil.Meter(mp)
+	counter, _ := meter.Float64Counter("command.time",
+		metric.WithDescription("Measures the duration of the buildx command."),
```
"build command" seems more appropriate as looks specific to that command atm.

tonistiigi commented on commands/build.go:
```diff
@@ -930,3 +1016,38 @@ func maybeJSONArray(v string) []string {
+
+// errorType returns an attribute for the error type based on the error category.
+// If nil, this function returns an invalid attribute.
+func errorType(err error) string {
```
`otelErrorType/metricsErrorType` maybe to make it clearer what "type" this is.

jsternberg commented on commands/build.go:
```diff
@@ -211,6 +219,88 @@ func (o *buildOptions) toDisplayMode() (progressui.DisplayMode, error) {
+func getVCSPath(contextPath string) string {
+	var wd string
+	if filepath.IsAbs(contextPath) {
+		wd = contextPath
```
The `getGitAttributes` function in `build/git.go` doesn't seem to handle this case. Does that need to be fixed too? It seems to get passed these parameters directly.

This code will also just fail and ignore the error if the git commands don't succeed. Is that suitable or do we need to catch the URL?

tonistiigi replied:
cc @crazy-max

jsternberg replied:
I've updated this to include a `Stat` call before the logic. That should catch any paths that don't make sense before we start to try running git commands.

crazy-max replied:
> `contextPath` in here can be `-` or a URL.

Hum right this was never checked when getting Git attributes, I'm looking at it.

tonistiigi commented on commands/build.go:
```diff
@@ -211,6 +219,61 @@ func (o *buildOptions) toDisplayMode() (progressui.DisplayMode, error) {
+func (o *buildOptionsHash) String() string {
+	o.resultOnce.Do(func() {
+		target := o.target
+		contextPath := o.contextPath
```
These paths need to be normalized so they always generate same hash(Eg. made absolute).

As this is leaking the system paths and Git info it still needs the salt like https://github.com/docker/buildx/pull/2185#discussion_r1465893251 . The multi-machine case doesn't apply anyway as system paths are different, and I don't think there is a way to implement it safely with some other way.

jsternberg replied:
Since this code is in a new place, I don't have easy access to the node identifier anymore. I'd have to make that function public.

jsternberg replied:
I've also now changed this so it will attempt to transform context path into a relative path. If a git directory is present, the relative path will be dependent on the root directory of the git repository.

tonistiigi commented on commands/build.go:
```diff
@@ -211,6 +219,61 @@ func (o *buildOptions) toDisplayMode() (progressui.DisplayMode, error) {
+func getGitRemoteURL(contextPath string) string {
+	// Check if this is a valid path to begin with before attempting
+	// to run git. It might be a non-filesystem URL.
+	if _, err := os.Stat(contextPath); err != nil {
```
I think we should do the URL/- detection to match `LoadInputs` and unless I'm missing something, same validation would need to be called for the `vcs` codepath.

If you think this is outside of scope, then create a follow-up issue for it and assign to yourself.

jsternberg replied:
I took a look at `LoadInputs` and it checks if the path is a local directory before checking for a URL. It seems to use this same code for that same functionality so I just exposed it and used that function instead. I've also included an explicit check for `-` to avoid attempting to call stat on that path and to avoid performing this logic at all in those circumstances.

We'll only do git remote url detection when it's a local directory.

crazy-max replied:
@jsternberg Looks good

crazy-max commented on build/utils.go:
```diff
@@ -34,7 +34,7 @@ func IsRemoteURL(c string) bool {
 }
 
-func isLocalDir(c string) bool {
+func IsLocalDir(c string) bool {
```
Maybe move this to `osutil`?

crazy-max commented on util/gitutil/gitutil.go:
```diff
@@ -196,3 +197,10 @@ func stripCredentials(s string) string {
 	return ru.String()
 }
+
+func ToAbs(path string) string {
+	if !filepath.IsAbs(path) {
+		path, _ = filepath.Abs(filepath.Join(osutil.GetWd(), path))
+	}
+	return SanitizePath(path)
+}
```
Could be moved to `osutil` as well but needs to move `SanitizePath` there too, can be for a follow-up.

jsternberg commented:
This PR is now also semi-dependent on this PR here: https://github.com/docker/cli/pull/4875

At the current moment, we don't have a way of obtaining the alias that was used to launch buildx. That PR sets an environment variable so the plugin can find the original command name so we can include it in the metric.

jsternberg commented:
I've removed the section of this that's dependent on `docker/cli`. That PR seems like it's likely going to come to either a different method of passing the data or setting something like `OTEL_SERVICE_NAME` and including this information as part of the service name.

I don't want to prevent this PR from being merged while we hash that out so I've removed it.

tonistiigi reviewed (changes requested):
@jsternberg Check the windows CI

jsternberg commented:
Was missing an import on the Windows build because my setup doesn't auto perform imports for systems that aren't compiled on my platform (aka windows in this case). Fixed it by adding the correct import. I'll keep an eye out because some stuff got moved around and that commonly means imports get removed and readded in new places so hopefully there's no more missing areas.
